### PR TITLE
Add data type check in ConvAddRelu fusion

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: 

The CUDA (and the CPU EP) only supports the float type for FusedConv.  When the ConvActivationFusion got split into SelectActionTransformers (https://github.com/microsoft/onnxruntime/commit/9e7d7a9e97c979c3b32fac4d54324e3f9c6b1d40#diff-e6b85539fe82c4b67655ba0f67140ad12548665d5633f820703e7c9cff0a87e2), the Conv+Add+Relu transformer didn't have the data type check duplicated in it - the Conv+Activation tranformer retained the original check before refactoring.

Additionally the data type check should also cover the CPU EP (https://github.com/microsoft/onnxruntime/pull/6653#discussion_r574958591)

**Motivation and Context**
Fixes https://github.com/microsoft/onnxruntime/issues/12045 and https://github.com/microsoft/onnxruntime/issues/11768
-This is a regression introduced in 1.11.

Fixes https://github.com/microsoft/onnxruntime/issues/11476  (No double type support in CPU FusedConv )

### CODE REVIEW

File: onnxruntime/core/optimizer/conv_activation_fusion.cc [hunk 1]
```diff
@@ -112,6 +127,10 @@ class ConvAddRelu : public NodeSelector {
       return std::nullopt;
     }
 
+    if (!ConvFusionDataTypeCheck(node)) {
```
Comment (hariharans29) [inline onnxruntime/core/optimizer/conv_activation_fusion.cc:130]: Check missing in ORT 1.11

File: onnxruntime/core/optimizer/conv_activation_fusion.cc [hunk 2]
```diff
@@ -74,12 +89,12 @@ class ConvActivation : public NodeSelector {
       return false;
     };
 
+    if (!ConvFusionDataTypeCheck(node)) {
+      return std::nullopt;
+    }
+
     // check EP type and activation
     if (node_ep == kCudaExecutionProvider) {
-      if (!HasElementDataType(*node.InputDefs()[0], ONNX_NAMESPACE::TensorProto_DataType_FLOAT)) {
```
Comment (hariharans29) [inline onnxruntime/core/optimizer/conv_activation_fusion.cc:79]: Moved out of the CUDA EP specific block

**Review by edgchen1 — DISMISSED** [review]
thanks!

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 3]
```diff
@@ -714,6 +714,36 @@ TEST_F(GraphTransformationTests, FuseCudaConvAddRelu) {
...
+TEST_F(GraphTransformationTests, FuseCudaConvAddRelu_UnsupportedType) {
+  auto model_uri = MODEL_FOLDER "fusion/conv_add_relu_fp16.onnx";
+  std::shared_ptr<Model> p_model;
+  ASSERT_STATUS_OK(Model::Load(model_uri, p_model, nullptr, *logger_));
+  Graph& graph = p_model->MainGraph();
+  for (auto& node : p_model->MainGraph().Nodes()) {
+    node.SetExecutionProviderType(kCudaExecutionProvider);
+  }
+  std::map<std::string, int> op_to_count = CountOpsInGraph(graph);
+  ASSERT_TRUE(op_to_count["Add"] == 1);
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/graph_transform_test.cc]: nit: consider ASSERT_EQ so both values are shown if they differ

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 4]
```diff
@@ -714,6 +714,36 @@ TEST_F(GraphTransformationTests, FuseCudaConvAddRelu) {
...
+// Currently the ConvAddRelu fusion is only backed by a float kernel for the
+// the CUDA EP.
+
+// When we see the corresponding pattern for the fp16 data type, the fusion
+// should not be triggered as there is no kernel to back the fused pattern.
+
+// TODO(hasesh): Limit the test to using the CUDA EP for now as the level of
+// data type support in other compatible EPs is still yet to be ascertained.
+
+// TODO(hasesh): If it all the fp16 type is supported for the fusion, adjust/remove
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/graph_transform_test.cc:726]: not quite following the comment
"If it all the fp16 type is supported for the fusion"

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/test/optimizer/graph_transform_test.cc]: Changed as part of fixing the Lint warnings, thanks

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 5]
```diff
@@ -714,6 +714,36 @@ TEST_F(GraphTransformationTests, FuseCudaConvAddRelu) {
...
+// this test.
+TEST_F(GraphTransformationTests, FuseCudaConvAddRelu_UnsupportedType) {
+  auto model_uri = MODEL_FOLDER "fusion/conv_add_relu_fp16.onnx";
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/graph_transform_test.cc:729]: would it be possible to add a model generation script too?

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/test/optimizer/graph_transform_test.cc:726]: That should read - "If at all,..." - does the comment make sense after that fix ?

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 5 above)
Comment (hariharans29) [reply onnxruntime/test/optimizer/graph_transform_test.cc:729]: Currently, we don't do add a model generation script for all tets models checked in. Also, in this case, I didn't generate the model. I used the existing fp32 model and chenaged the data types. Can this be addressed later ?

Comment (edgchen1) [reply onnxruntime/test/optimizer/graph_transform_test.cc:729]: sure

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**